### ISSUE TITLE
Integrate FSRS into Anki as an optional feature

### ISSUE THREAD

**Issue by dae** [ankitects/anki#2443]
Currently users need to paste code into the custom scheduling section of the deck options in order to use FSRS. It may make sense to integrate this code directly into Anki, to make it easier to use.

According to Jarrett, the scheduling code does not change frequently: https://github.com/ankitects/anki/pull/2421#issuecomment-1457925875

Users may wish to apply different weights to different decks. The existing custom scheduling code relies on deck name matches (prefix matches?) in order to determine the weights. We could perhaps store the weights in the deck presets - this would allow a deck and its subdecks to share the same weights. Alternatively the weights could be stored in individual decks, which is the most flexible, but will lead to extra storage. The weights do not appear to be particularly large: https://github.com/ankitects/anki/pull/2421#issuecomment-1460046666

This would be an optional feature, ~and would require the add-on/weights trained externally for now. If it is successful, it may make sense to look into a Rust-based training method in the future, so that users don't need to rely on external sites/programs~.

@L-M-Sherlock @RumovZ any thoughts?

**Comment by L-M-Sherlock** [comment]
I would like to know where the code of FSRS could be integrated into Anki. FSRS only modifies the `customData` and the `scheduledDays` of `ReviewState`.

**Comment by dae** [comment]
It would probably require changes to states/review.rs to apply the FSRS scheduling if an option had been set; the actual FSRS code would best be placed in a new file.

**Comment by L-M-Sherlock** [comment]
Here are some codes I think are related:

https://github.com/ankitects/anki/blob/146bed12d9e14bee8620b61f367bd95f61b7d82e/rslib/src/scheduler/states/learning.rs#L53

https://github.com/ankitects/anki/blob/146bed12d9e14bee8620b61f367bd95f61b7d82e/rslib/src/scheduler/states/learning.rs#L70

https://github.com/ankitects/anki/blob/146bed12d9e14bee8620b61f367bd95f61b7d82e/rslib/src/scheduler/states/learning.rs#L80

https://github.com/ankitects/anki/blob/146bed12d9e14bee8620b61f367bd95f61b7d82e/rslib/src/scheduler/states/relearning.rs#L45

https://github.com/ankitects/anki/blob/146bed12d9e14bee8620b61f367bd95f61b7d82e/rslib/src/scheduler/states/review.rs#L59

https://github.com/ankitects/anki/blob/146bed12d9e14bee8620b61f367bd95f61b7d82e/rslib/src/scheduler/states/review.rs#L80

If FSRS is enabled, the scheduler will call a new function in a file, maybe named fsrs.rs.

**Comment by dae** [comment]
Do you have thoughts on whether we should put the weights in deck presets or individual decks?

**Comment by L-M-Sherlock** [comment]
> Do you have thoughts on whether we should put the weights in deck presets or individual decks?

Deck presets are better. The current built-in algorithm also reads its configurations from deck presets.

**Comment by L-M-Sherlock** [comment]
> it may make sense to look into a Rust-based training method in the future, so that users don't need to rely on external sites/programs.

My friend and I have tried [tch](https://docs.rs/crate/tch/latest), which provides Rust bindings for the C++ api of PyTorch. But it required [libtorch](https://pytorch.org/cppdocs/installing.html), whose size is about 200MB.

**Comment by dae** [comment]
Yeah, that's too large I'm afraid. If it were to get integrated into Anki, it would need to be a smaller library, and preferably one that's implemented in native Rust. I don't know much about machine learning, so I don't know what sort of things your code relies on, or how easy/practical it would be to port the existing code to one of the options listed on https://www.arewelearningyet.com/

**Comment by L-M-Sherlock** [comment]
I have tried a small library, tinygrad. Now we can get rid of torch: [open-spaced-repetition/fsrs-optimizer-tiny (github.com)](https://github.com/open-spaced-repetition/fsrs-optimizer-tiny). It only requires numpy and pandas (and we also can find a lightweight alternative for pandas).

However, tinygrad is Python-based and still in alpha. There is still a long way to go.

**Comment by dae** [comment]
Sorry, I wasn't clear before - when I mentioned native Rust, I meant rather than a Rust wrapper for a C/C++ library. Python-based libraries aren't really an option, as any code added to Anki's core needs to work on the mobile clients too.

**Comment by L-M-Sherlock** [comment]
> Rust wrapper for a C/C++ library

Most of them are bindings for Tensorflow and PyTorch. So their size is very large. Most naive Rust frameworks are experimental.

**Comment by user1823** [comment]
If the devs are successful in incorporating the optimizer into Anki, it would be undoubtedly a huge achievement.

But, for now, I think that the devs should focus on incorporating the scheduler and the features from the helper add-on into Anki.

If the scheduler and helper add-on code is incorporated into Anki, it would
- make it much easier to configure the scheduler. Currently, many users do mistakes in putting the values of the parameters (w) in the scheduler code and also many of them struggle with configuring it differently for different decks.
- enable things like creating filtered decks based on FSRS properties.

Also, given that training is something that the user doesn't need to do very frequently (only once a month, as L-M-Sherlock suggests), using an external site/program for it would not be very inconvenient.

**Comment by L-M-Sherlock** [comment]
I agree with @user1823. @dae, my friend and I implement a Rust version for FSRS scheduler, here is the repo: https://github.com/open-spaced-repetition/rs-fsrs. Maybe it would be helpful.

**Comment by dae** [comment]
To summarize what is required here, as I currently understand it:

- We add a field to DeckConfig to store the vector of weights, which should be a vector of at least a certain length.
- We introduce a setting in DeckConfig that controls whether FSRS is on (or maybe better: use the presence/absence of the weights?)
- We update scheduler/states/review.rs:next_states() to provide alternate values when FSRS is enabled, based on the logic in https://github.com/open-spaced-repetition/rs-fsrs. Might be nice to keep the actual FSRS scheduling logic in a separate file or files.

I presume that's the minimum required to enable usage of FSRS without using custom scheduling. The existing add-on could write the weights into the deck config, until we are able to integrate the weight calculation inside Anki directly.

**Comment by L-M-Sherlock** [comment]
> * We add a field to DeckConfig to store the vector of weights, which should be a vector of at least a certain length.

And request retention. Here is typical configuration for FSRS:

```json
{
    "deckName": "global config for FSRS4Anki",
    "w": [0.4, 0.6, 2.4, 5.8, 4.93, 0.94, 0.86, 0.01, 1.49, 0.14, 0.94, 2.18, 0.05, 0.34, 1.26, 0.29, 2.61],
    "requestRetention": 0.9,
    "maximumInterval": 36500,
  }
```

The "deckName" and "maximumInterval" can be removed when we enable usage of FSRS without using custom scheduling.

**Comment by RumovZ** [comment]
> use the presence/absence of the weights

That would mean you'd lose the weights after toggling FSRS off and on again, though, as far I understand, which is a bit inconvenient.

I assume the card-specific parameters would remain in `custom_data`?

**Comment by dae** [comment]
> That would mean you'd lose the weights

Yep, good point.

> would remain in custom_data

Yeah, I think that's probably best for now, as a DB schema change will be moderately expensive, and would require changes to the sync protocol.

Of the parameters shown in the screenshot here:

https://forums.ankiweb.net/t/how-to-use-the-next-generation-spaced-repetition-algorithm-fsrs-on-anki/25415/223?u=dae

- I presume v=helper has been used to identify what wrote the details; maybe we could shorten this to something like v=1 for Anki, or omit it entirely?
- @L-M-Sherlock what is seed used for? What are its min/max values, and would it be problematic if we used a shorter name?

**Comment by L-M-Sherlock** [comment]
> * I presume v=helper has been used to identify what wrote the details; maybe we could shorten this to something like v=1 for Anki, or omit it entirely?

`v` denotes version. I store it just to locate the problem, because it has two source: the schedule and the helper. If we will implement `reschedule` feature, it would be useful to store the source.

> * what is seed used for? What are its min/max values, and would it be problematic if we used a shorter name?

The seed is fuzz factor. The reason I store it is that the random generator is different between JavaScript and Python. I need to store the fuzz factor to keep consistent between the scheduler (js script) and the helper (py add-on).

**Comment by Expertium** [comment]
Suggestion: allow users to have different parameters for different decks **or** apply the same parameters to all decks.

**Comment by dae** [comment]
That's the plan. https://github.com/ankitects/anki/issues/2443#issuecomment-1474743194

**Comment by user1823** [comment]
Congratulations for successfully integrating the FSRS optimizer into Anki (https://github.com/ankitects/anki/pull/2633).

For the UX, I have a suggestion.
- Create a new option at the top of the Deck Options page for enabling/disabling FSRS.
- Use the state of this option to decide whether to display specific deck options or not. For example, 
    - Don't display the `Optimize` option when FSRS is disabled.
    - Don't display `graduating interval`, `easy interval`, `new interval`, `starting ease`, `interval modifier`, `easy bonus`, and `hard interval` when FSRS is enabled.

Also, the user should be allowed to set a `requestRetention` other than the one suggested by the optimizer. The suggested `requestRetention` is aimed at minimizing the workload but the user must be allowed to adjust the scheduler so as to achieve their learning goals (as is possible via the custom scheduling version of FSRS).

Lastly, `Check` should be renamed to something like `Evaluate`.

**Comment by Expertium** [comment]
This might be a bit too much to ask, but I think the Supermemo approach of choosing the requested R (in Supermemo, "fogetting index" is just one minus requested retention) via a slider is really good. It ensures that the user cannot enter non-numerical characters and also ensures that the user cannot set the requested retention (or forgetting index) higher or lower than certain values. In SuperMemo, the forgetting index cannot be set lower than 3% (requested retention 97%) or higher than 20% (requested retention 80%), and I think that's a perfectly reasonable range and we should do the same.
![image](https://github.com/ankitects/anki/assets/83031600/9759fa23-1d8f-4730-a4d2-6b2d88cf2164)

**Comment by user1823** [comment]
> I think the SuperMemo approach of choosing the requested R via a slider is really good. … In SuperMemo, the forgetting index cannot be set lower than 3% (requested retention 97%) or higher than 20% (requested retention 80%), and I think that's a perfectly reasonable range and we should do the same.

I agree that we should only allow requestRetention values in the range 80% to 97%. But I don't like the idea of using a slider because using a slider to set a value is less convenient than simply typing it out.

**Comment by dae** [comment]
> Create a new option near the top of the Deck Options page for enabling/disabling FSRS.

A toggle already exists in the advanced section. It would definitely make sense to hide certain options when it is enabled, or display warnings (eg if a learning step is longer than a day and fsrs is enabled, warn the user that it is not advised).

The requested retention is editable by the user, and the plan is to make it use a spinbox like the other number fields, with appropriate min and max limits. I'm not sure we really need a slider in addition to a spinbox.

**Comment by Expertium** [comment]
Another suggestion: if it's possible to run the optimizer on each individual deck, then it should be possible to run it only if the deck has a certain number of reviews (say, 1000), otherwise, an error message should pop up, telling the user that the data is insufficient to find optimal parameters accurately, and then the parameters of the parent deck should be used instead.

**Comment by dae** [comment]
We should definitely add a warning if there's not enough source material, but revlogs are gathered based on preset, not deck. Advanced users who want to use some custom training scope can use the search, and it could be easily automated in an add-on.

**Comment by Expertium** [comment]
A minor suggestion: currently if users want to check whether FSRS is working, they have to modify `const display_memory_state = false` in the scheduler code. Once FSRS is integrated into Anki, there should be a button to toggle this on/off.

**Comment by user1823** [comment]
> Once FSRS is integrated into Anki, there should be a button to toggle this on/off.

This is not required in my opinion. With the custom scheduling approach, it makes sense to add this option because the implementation is fragile. But, when FSRS is natively integrated into Anki, it is not required because you can trust that the scheduler is working.

However, that option also displayed the memory states (D, S, R). This might be worth adding (to the Card Info). It already contains D and S in the Custom Data but R is not there. Another approach could be to add these memory states to the Browser (as the Helper add-on currently does).

**Comment by Expertium** [comment]
> but R is not there

The problem is that S and D don't change in between repetitions, but R does.

**Comment by dae** [comment]
The plan is to add DSR to the card info and browser. I've already done the card info, and it uses the current R, not R at last review.

**Comment by Expertium** [comment]
> It would definitely make sense to hide certain options when it is enabled

Speaking of which, aside from hiding settings, it also makes sense to hide stats related to card's Ease once FSRS is enabled.

**Comment by L-M-Sherlock** [comment]
<img width="305" alt="image" src="https://github.com/ankitects/anki/assets/32575846/e0eb6827-2f87-4b70-807c-a38af1b7c563">

These values can be extracted from revlog. If we implement auto-filling, I recommend folding these fields to avoid confusing the user.

How about opening a new issue to discuss about that?

**Comment by dae** [comment]
I've opened https://github.com/ankitects/anki/issues/2661

You're welcome to open issues for any other problems you notice, too - please give them the FSRS tag.

**Comment by Expertium** [comment]
I know that this issue is already closed, but still, I want to ask something: do we really need (re)learning steps with FSRS? I don't see why we would keep them.
1) It can cause problems such as "Good" interval being shorter than "Hard" interval.
2) It adds another setting for users to worry about and to (needlessly) tweak, and we want to have as few settings as possible to decrease the cognitive load for the users and make everything simple and straightforward.
3) FSRS already has 4 parameters for initial stability after only one review, we could just use those.
In other words, the (re)learning stage can be removed entirely when FSRS is enabled. Of course, (re)learning steps will still be available to those who stick with the old algorithm.

**Comment by user1823** [comment]
> do we really need (re)learning steps with FSRS? I don't see why we would keep them.

The reasons (re)learning stage is required:
- If the user presses Again on a new card, they need to see it again on the same day so that they can memorize it. If we remove the learning steps, the shortest interval the card can get is 1 day. This would not be acceptable to most users. The same holds for the review cards that were rated Again.
- The initial stability depends upon the learning steps. If the (intraday) learning steps are changed significantly, there can be a significant change in the initial stability. This means that the initial stability trained using old reviews will not match the initial stability of the new reviews.

**Comment by Expertium** [comment]
Regarding your first point: Woz says that same-day reviews have a negligible impact on long-term memory, and FSRS indirectly supports it as well; the optimizer works just fine when same-day reviews are removed. So we could just tell the users "no more same-day reviews".
Regarding your second point: yes, the optimal value of initial S can change, but then it can just be re-estimated as new data (the user is doing reviews) comes in.

**Comment by user1823** [comment]
> Woz says that same-day reviews have a negligible impact on long-term memory

[Here](https://supermemo.guru/wiki/Myth:_Review_your_material_on_the_first_day_several_times), Wozniak says:
> A single effective review on the first day of learning is all you need. Naturally, it may happen you cannot recall a piece of information upon a single exposure. In such cases, you may need to repeat the drill. 

So, if the user presses `Again`, the review should be repeated on the same day.

In addition, I think that intraday reviews (of failed cards) can be helpful in making an early segregation of cards based on their difficulty. However, the current version of FSRS completely misses this opportunity by filtering out intraday reviews. See [this issue](https://github.com/open-spaced-repetition/fsrs4anki/issues/426) for details.

> yes, the optimal value of initial S can change, but then it can just be re-estimated as new data (the user is doing reviews) comes in.

But then, the optimized value of initial S would match neither the initial S of the old cards nor that of the new cards. It would rather be in between these two values.

**Comment by Expertium** [comment]
> So, if the user presses `Again`, the review should be repeated on the same day.

Then how about making learning steps the same for everyone (15m 1d, for example) - and hiding them aka removing this setting when FSRS is enabled? We already know that long learning steps can cause problems with FSRS, so it makes sense to not allow users to tweak them when FSRS is enabled.

**Comment by user1823** [comment]
> Then how about making learning steps the same for everyone

It makes sense. I also thought about doing the same. The only problem is that users might complain that they have been robbed of the ability to adjust the learning steps.

> 15m 1d, for example

I recommend using only `15m`. FSRS can control the long-term scheduling much better.

I don't remember why, but about 2.8 years ago, I set my first learning step as `15m` and my first relearning step as `20m`. These (re)learning steps have produced great results since then. So, I recommend using these steps if we intend to use same steps for everyone.

**Comment by Expertium** [comment]
@dae @L-M-Sherlock what do you think? My philosophy is that the user should have to tweak as few settings as possible. If something _can_ be done by the software without the user's intervention, then it _should_ be done by the software.

**Comment by dae** [comment]
You're saying that 15 minutes works best for you, and suggesting we force it on everyone. I'm not sure that's a good idea, and suspect people with different preferences would not be fond of such a change. FSRS can get away with fewer user knobs because the weights are trained based on individual users' data, but a hard-coded learning step wouldn't be.

> It can cause problems such as "Good" interval being shorter than "Hard" interval.

Anki shows a warning when the user has FSRS enabled and sets a (re)learning step over a day, and if they set large steps anyway, they should be handled correctly, as the elapsed days between learning steps is fed into FSRS.

**Comment by dae** [comment]
Perhaps an alternative is to increase the guard rails - eg showing a warning if the user adds more than 2 steps/1 relearning step, or steps over an hour. This nudges people towards sensible choices, without forcing them or hard-coding a particular value.

**Comment by user1823** [comment]
> You're saying that 15 minutes works best for you, and suggesting we force it on everyone. I'm not sure that's a good idea,

I think that you misunderstood my point (though it's not your fault, I wasn't clear enough). It is not that I played around with different settings of learning steps and then settled on one value that best suits me. In that case, I wouldn't have recommended using that value for all users.
The thing is that I read this value being recommended somewhere online and after I set it, my learning improved significantly. If the value recommended by someone else (without knowing the contents of my cards and my learning ability) works great for me, it should work well for others too.

> Perhaps an alternative is to increase the guard rails - eg showing a warning if the user adds more than 2 steps/1 relearning step, or steps over an hour. This nudges people towards sensible choices, without forcing them or hard-coding a particular value.

This seems to be a reasonable alternative.

**Comment by Expertium** [comment]
I want to add that the point is not to find the perfect value of learning steps that works great for everyone, the point is to set it to _some_ value that is _good enough_ and "lock" it to prevent:

1) Users worrying "How should I tweak this setting?"

2) Problems with FSRS and long learning steps

**Comment by L-M-Sherlock** [comment]
According to my research in my job, learning steps really affect the long-term memory. Long learning steps would increase the memory stability, but it also increase the error rate during intraday learning.

> In Nakata (2015a), for instance, 128 Japanese university students studied 20 English-Japanese word pairs using flash card software. The participants were randomly assigned to one of the following four spacings: massed-, short-, medium-, and long-spaced. The participants in the massed group practiced retrieval of a given English word three times in a row. The short-, medium-, and long-spaced groups used average spacing of approximately one, two, and six minutes, respectively (that is, participants in the long-spaced group practiced retrieval of a given target word approximately every six minutes). Nakata found that the massed group was the least effective among the four groups both immediately and one week after the treatment, supporting the spacing effect. Moreover, while the short-spaced group fared significantly bet- ter than the massed group immediately after learning, the advantage disappeared at a one-week retention interval. The medium- and long-spaced groups, in contrast, were significantly more effective than the massed group not only on the immediate but also on the one-week delayed posttest. The findings are consistent with the spacing-by-retention interval interaction, which states that the benefits of short lags disappear over time. Nakata and Webb (2016a, Experiment 2) also observed the positive effects of long lags for L2 vocabulary learning. They compared the effects of short and long spacing. Encounters of a given target item were separated by approximately 30 seconds and three minutes in the short- and long-spaced conditions, respectively. While the long-spaced group failed to outperform the short-spaced group immediately after the treatment (short = 58%; long = 62%), the long-spaced group fared significantly better than the short-spaced group at a one-week retention interval (short = 8%; long = 20%; Figure 20.1). These findings support the lag effect and spacing-by-retention interval interaction.

Source: https://www.taylorfrancis.com/chapters/edit/10.4324/9780429291586-20/learning-words-flash-cards-word-cards-tatsuya-nakata

**Comment by user1823** [comment]
> According to my research in my job, learning steps really affect the long-term memory.

Yes, they definitely do. But, in my opinion, the only thing that matters is that the time gap between a successful review and the previous failed review.

The reason for this is that when the time gap is too small, there is no "spacing effect" and thus, the review doesn't cause memory stabilization.

However, adding multiple steps on the same day doesn't affect the long-term memory.

**Comment by L-M-Sherlock** [comment]
@dae, I'm refactoring my add-on, and find out a problem. What if the user do reviews in other devices which haven't supported FSRS? Should we re-calculate the memory state for these cards synced from other devices?

**Comment by Expertium** [comment]
Out of curiosity, can someone tell me how long (roughly) it will take before a new version of Anki, with FSRS built-in, is released? I know it's hard to say, but I just want to know whether it's more like 2 weeks or many months.

**Comment by dae** [comment]
I suggest we initially put a warning in the UI when people enable FSRS: "For this feature to work properly, all of your Anki clients must be Anki(Mobile) 23.10+ or AnkiDroid 2.17+". I [submitted some PRs](https://github.com/ankidroid/Anki-Android/pull/14456) to AnkiDroid yesterday which adds support for the new FSRS functionality.

The first step to a release is a beta, I hope to do that within a week.

**Comment by dermpathMD** [comment]
In your opinion, is it fine to install FSRS 4 on my own now or should I wait for the beta with FRSR integrated? Or does it not really matter?

**Comment by meliache** [comment]
I'm curious to try the Beta soon. Would it be possible to set the *review sort order* to the probability of recall $R$? For working of large backlogs there is the *relative overdueness* sort option, but it's a totally arbitrary metric which aims to show cards that are more likely to be forgotten first. With FSRS I think that sorting option does not make much sense and it's better to utilize the memory-state knowledge of FSRS. If this is not already included somewhere or planned anyway I might formulate a fully-fledged PR.

I'm just checking if I can create a filtered deck bared on $R$ (or reschedule cards with a good $R$), which would be an alternative workaround for working through large backlogs, but I'm not sure if I can query for $R$, as it's a calculated property. Querying for the stability $S$ works though nicely via `prop:cdn:s`, and this is good enough. Conceptually this seems an alternative way to detect "mature" cards to reschedule, if I understand correctly.

**Comment by Expertium** [comment]
@dae I like the idea proposed above. Simply put, add "Ascending R" and "Descending R" to the "Review sort order" menu. Also, remove "Ascending ease" and "Descending ease" if FSRS is enabled.

**Comment by dae** [comment]
I like the idea, but I think it might be problematic when we enable some decks to use SM-2 and some to use FSRS, as the sorting will be incorrect when there's a mix of SM-2 and FSRS decks. Perhaps we should be making FSRS a global option instead of per deck-preset? The weights would still be per-preset.

**Comment by Expertium** [comment]
> Perhaps we should be making FSRS a global option instead of per deck-preset? The weights would still be per-preset.

I thought that's the plan.

**Comment by user1823** [comment]
> I like the idea, but I think it might be problematic when we enable some decks to use SM-2 and some to use FSRS, as the sorting will be incorrect when there's a mix of SM-2 and FSRS decks. Perhaps we should be making FSRS a global option instead of per deck-preset?

In my opinion, this is an issue only when the user has configured overlapping presets such that the parent deck is assigned one preset (where FSRS is enabled), and the subdeck is assigned another preset (where FSRS is disabled).

Making FSRS a global option will work. But for the current issue, I think that it is sufficient to display a warning while setting "Review Sort Order" to Ascending/Descending R if any of the subdecks of the decks affected by the preset have FSRS disabled.

Thinking more about it, the decks affected by a preset can also be changed later. So, the above warning should also be displayed if the user changes the preset used by a deck and this change causes the above problem.

Obviously, all this can be prevented just by making FSRS a global option. But, deciding whether to do this or not depends upon finding how common it for users is to enable FSRS for specific decks only.

**Comment by user1823** [comment]
I wanted to ask a question related to presets and preset-specific weights.

Suppose that I have one parent deck that has one preset. Then, I have several subdecks, each with different presets. Can I use the same weights for all these decks? In other words, is it possible to have subdecks use weights of the parent deck even if they have different presets?

The reason for using different presets for subdecks is to control the number of new cards I get from each of them. But, in terms of their difficulty, they have similar cards. Evidence suggests that FSRS works better if you use one set of weights for all the cards rather than several sets of weights if the cards are similar in terms of their difficulty.

**Comment by user1823** [comment]
> add "Ascending R" and "Descending R" to the "Review sort order" menu

With this suggestion, I remember the discussion that happened when developing the "Advance" and "Postpone" features in the FSRS Helper add-on.

Sorting cards based only on the retention (R) creates problems because different cards may have different request retention. A card may have low R, but this card might not be a priority because it might have low request retention as well. So, it might be better to use the sorting formulas used in postpone.py and advance.py of the helper add-on.

https://github.com/open-spaced-repetition/fsrs4anki-helper/blob/e709cd48910155814a9c212be741afbfd11fb761/schedule/postpone.py#L94-L96

https://github.com/open-spaced-repetition/fsrs4anki-helper/blob/e709cd48910155814a9c212be741afbfd11fb761/schedule/advance.py#L95-L97

**Comment by Expertium** [comment]
I think that's needlessly complicated, and users may start asking whether sorting is bugged if they are shown a card with a low R, then a card with a higher R, and then a card with a lower R again.

**Comment by user1823** [comment]
> users may start asking whether sorting is bugged

But I am not suggesting that we should name this new option as Ascending/Descending R. We can continue using the current name (Relative Overdueness) or a similar name because the concept is the same, just the calculation is more accurate. Here, instead of using the scheduled interval (which includes fuzz), we use the calculated interval (which doesn't include fuzz).

If you didn't get what I said, let me tell you that these values are related as follows:
$$\frac{\frac{1}{R} - 1}{\frac{1}{R_0} - 1} = \frac{\text{Elapsed days}}{\text{Calculated interval}}$$

Here, R = current interval and R0 = request retention

**Comment by Expertium** [comment]
If it's just a different way of calculating overdueness, then I guess it's fine. @meliache what do you think?

**Comment by meliache** [comment]
@Expertium I kept out of the discussion because much more informed people got involved. I was not aware of the equivalency that @user1823 mentioned, I didn't think that with FSRS relative overdueness sorting and R sorting (given a fixed $R_0$) are equivalent.

Not using fuzz in the calculation seems to make sense. So I now agree we should keep the relative overdueness sorting option by name as-is and just consider improving the calculation based on the FSRS-calculated interval instead of the schedulead interval (or equivalently via this $\frac{1/R-1}{1/R_0-1}$ formula).

Beyond fuzz another difference between the two sides of the equation above can arise when the FSRS deck contains cards where the interval was not scheduled by FSRS. With the custom scheduling method only new cards are rescheduled and with the FSRS helper there is the option to only reschedule cards studied in the last 7 days. So currently you can have a deck where old cards still use non-FSRS intervals. Not sure how that will be when FSRS is in the core.

Also of course there is the option to manually set the due-date of a card, should that affect the "relative overdueness"? It would be weird if that action affects the "relative overdueness" sorting for non-FSRS cards but not for FSRS cards. Therefore if possible I would prefer sorting based on a relative overdueness based on the ideal calculated (not actually scheduled) interval for all cards, even non-FSRS ones. If we do that, of course that should be documented in the info popup card for the sorting options.

**Comment by dae** [comment]
> Suppose that I have one parent deck that has one preset. Then, I have several subdecks, each with different presets. Can I use the same weights for all these decks? In other words, is it possible to have subdecks use weights of the parent deck even if they have different presets?

With the v3 scheduler, you can use the same preset for all of your decks, and give each deck its own daily limit. If you wished to keep using separate presets, you'd need to copy the weights into each preset, and use a custom search when computing them.

**Comment by Expertium** [comment]
> Also of course there is the option to manually set the due-date of a card, should that affect the "relative overdueness"?

@L-M-Sherlock what do you think? If we change sorting by relative overdueness so that it uses current R and requested R when FSRS is enabled, how should this be handled?

**Comment by L-M-Sherlock** [comment]
I have provided the sorting formula in the helper add-on:

https://github.com/open-spaced-repetition/fsrs4anki-helper/blob/e709cd48910155814a9c212be741afbfd11fb761/schedule/postpone.py#L64-L99

**Comment by user1823** [comment]
I was trying the Beta release and noticed two issues:
1. The weights generated by Anki are very different from those generated by the python optimizer.
- Weights generated by Anki: 1.1408, 2.1175, 9.0984, 90.9615, 4.7957, 1.1697, 1.2771, 0.0010, 1.6411, 0.1000, 1.0769, 2.1627, 0.0928, 0.4849, 1.0269, 0.0000, 3.4083
- Weights generated by python optimizer: 1.1366, 1.4954, 20.8954, 41.342, 4.5015, 1.6955, 2.114, 0.0, 1.7707, 0.1551, 1.1898, 1.5944, 0.1889, 0.6854, 0.0115, 0.0, 6.6447
- Deck to reproduce: 
[Default.zip](https://github.com/ankitects/anki/files/12708324/Default.zip) (change file extension to .apkg)
- Timezone (for Python optimizer): Asia/Calcutta
- Other settings are Default
This issue is not caused by the bug where Anki calculates weights for all the reviews (not respecting the preset).
2. Anki doesn't seem to reschedule the cards after updating the weights and/or desired retention.

**Comment by Expertium** [comment]
Wait, beta is already released? Can you give me a link?
Also, it's probably better to create a new issue for everything related to FSRS and the beta version

**Comment by user1823** [comment]
> Wait, beta is already released? Can you give me a link?

https://github.com/ankitects/anki/releases/tag/23.10beta1

**Comment by dae** [comment]
Re: rescheduling - I wonder if that's the best idea? The advantage of the current approach is that users can enable FSRS without penalty. I don't want them to enable FSRS and then suddenly have hundreds more reviews to do - I feel like that should be some separate operation.

**Comment by user1823** [comment]
> I feel like that should be some separate operation.

I agree.

**Comment by Expertium** [comment]
Here's my feedback so far
1) Maximum interval and desired retention do not have descriptions, even though hovering the mouse over them shows a question mark (left clicking on them does nothing)
![image](https://github.com/ankitects/anki/assets/83031600/a21ecd2f-56ea-4e03-83dd-92fe00450f5c)
2) The number of reviews is the smae for nay preset, meaning that the optimizer doesn't actually work on a per-preset basis, and goes through the entire collection instead.
![image](https://github.com/ankitects/anki/assets/83031600/5d958394-8f7b-44e1-9c78-b9a5b19263f7)

3) This one is especially surprising: I get absolutely horrendous mouse stuttering (I'm not sure how else to describe it) in this Anki version. I have a 144 hz monitor and it feels like 30 hz when I'm moving my mouse.

**Comment by user1823** [comment]
After clicking Compute with this search string, Anki automatically entered these weights.

<img width="510" alt="image" src="https://github.com/ankitects/anki/assets/92206575/48f62925-92e6-4edd-b9c2-bdea6d7948f2">

These weights don't seem to be the [default weights](https://github.com/open-spaced-repetition/fsrs4anki/blob/5a34201f4d5b3a9cc1a5da4511b6bc32b7c6e909/fsrs4anki_scheduler.js#L11). So, are these weights calculated from this one card only? If so, I suggest entering the default weights (instead of the computed weights) when the number of reviews is small.

**Comment by L-M-Sherlock** [comment]
> 2\. The number of reviews is the same for any preset, meaning that the optimizer doesn't actually work on a per-preset basis, and goes through the entire collection instead.

You need to input query for collecting the training review logs. Such as deck:"xxxx".

**Comment by user1823** [comment]
I have some more suggestions (for the UX):
- show a popup when the user sets desired retention beyond the permissible range. "The permissible range for desired retention is 0.80 to 0.97." Currently, the value changes without any feedback. This seems to be a poor UX.
- The same goes for calculating optimal retention. Show a tooltip saying, "The calculated optimal retention has been entered into the desired retention box." 
- Remove the 100.00% that remains below the compute button. I initially thought that the calculated optimal retention was 100% for me. If you don't want to remove it, change it to "XX.XX% complete."

Also, a question: How to search for memory states in the browser?

> The number of reviews is the same for any preset, meaning that the optimizer doesn't actually work on a per-preset basis, and goes through the entire collection instead.

@Expertium, this bug has been fixed and the next beta should work fine. Use preset:"Preset name" until then.

**Comment by L-M-Sherlock** [comment]
> 4\. As user8123, there is no option to reschedule all cards.

It's automatic. When you modify the weights and click `save`, Anki will re-calculate all cards' memory states.

**Comment by user1823** [comment]
> It's automatic. When you modify the weights and click save, Anki will re-calculate all cards' memory states.

It re-calculates only the memory states, not the intervals and due dates.

**Comment by Expertium** [comment]
> You need to input query for collecting the training review logs. Such as deck:"xxxx".

That is very confusing, and not all users will know the syntax for that. It would be much better to just make the optimizer work on the selected preset.

**Comment by L-M-Sherlock** [comment]
> Also, a question: How to search for memory states in the browser?

For example: `prop:s<5`, `prop:d>3`, `prop:r<0.5`.

> not the intervals and due dates.

Yeah, but it wouldn't induce backlog, like auto-reschedule after each review.

**Comment by Expertium** [comment]
Question: does this version use online learning, or do you need to manually re-run the optimizer occasionally?

**Comment by L-M-Sherlock** [comment]
By the way, this issue has been off-topic. I recommend posting your question and feature request in the forum: https://forums.ankiweb.net/

**Comment by L-M-Sherlock** [comment]
> Question: does this version use online learning, or do you need to manually re-run the optimizer occasionally?

It still require user to re-run the optimizer occasionally.

**Comment by dae** [comment]
>  I get absolutely horrendous mouse stuttering

Perhaps changing the video driver in the preferences screen will help. If not, try the qt5 version.
